### PR TITLE
Clean up reference domain dependency in gmem consumer indexing

### PR DESCRIPTION
**Pull request by shmsong**

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp [hunk 1]
```diff
@@ -14,6 +14,39 @@ namespace cuda {
...
+    if (Index::protectWithMagicZero(loop, concrete_loop_id, loop_ind)) {
+      magic_zero_loop = concrete_loop_id;
+    }
+  }
+
+  // Insert magic zero if insertion point found
+  if (magic_zero_loop != nullptr &&
+      concrete_loop_idx_map.count(magic_zero_loop)) {
+    auto& ind = concrete_loop_idx_map.at(magic_zero_loop);
+    if (!ind->isConstScalar()) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp:51]: Should this check be part of trying to find an insertion point? If an inner loop is a const scalar index, but a loop in a more outer position wasn't, shouldn't we just put it on the loop that's in a more outer position?

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp:51]: I would presume when we have more optimal index hoisting we shouldn't need magic zero anymore 🤞

**Review by csarofeen — APPROVED** [review]
LGTM, minor comment on magic zero.

File: torch/csrc/jit/codegen/cuda/lower_index_hoist.h [hunk 2]
```diff
@@ -96,6 +110,16 @@ class TORCH_CUDA_CU_API CommonIndexMap {
       const std::vector<kir::ForLoop*>& loops,
       Val* index);
 
+  //! Duplicate of above, but without a reference domain. TODO: Remove other
+  //! implementation.
+  std::pair<Val*, bool> insert(
+      IterDomain* indexed_consumer_id,
+      TensorDomain* consumer_td,
+      std::vector<IterDomain*> loop_domains,
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_index_hoist.h]:
```suggestion
      const std::vector<IterDomain*>& loop_domains,
```

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 3]
```diff
@@ -1167,6 +1184,40 @@ Val* hoistConsumerIndex(
...
 }
 
+// Version of hoisting without using reference tensor,
+//  should eventually deprecate the other one once reference
+//  tensor is completely deprecated.
+Val* hoistConsumerIndex(
+    IterDomain* consumer_root_id,
+    const TensorView* consumer_tv,
+    const IndexCompute& consumer_indexing,
+    std::vector<IterDomain*> loop_domains,
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]:
```suggestion
    const std::vector<IterDomain*>& loop_domains,
```

File: torch/csrc/jit/codegen/cuda/lower_index_hoist.h [hunk 4]
```diff
@@ -43,6 +43,20 @@ class CommonIndexKey {
...
+  //! \param consumer_td TensorDomain of consumer_indexed_id
+  //! \param loop_domains Resolved vector of iterdomain corresponding to loops
+  //! \param loop_index_map Index mapping generated from the loop nest.
+  //! \param loops Loop structure where this id is indexed
+  //! Duplicate of above, but without a reference domain. TODO: Remove other
+  //! implementation.
+  CommonIndexKey(
+      IterDomain* consumer_indexed_id,
+      TensorDomain* consumer_td,
+      std::vector<IterDomain*> loop_domains,
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_index_hoist.h]:
```suggestion
      const std::vector<IterDomain*>& loop_domains,
```

File: torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp:51]: As far as I remember, manually hoisting wasn't helpful as the compiler was able to discover the "opportunity" to cache indices.

**Review by naoyam — APPROVED** [review]
LGTM as well. Just some minor comments.

File: torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp (hunk 1 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp:51]: Would need to revisit magic zeros in follow ups.

**Pull request merged**